functions/documents.ts

```diff
@@ -1,25 +1,465 @@
+  ? normalizeTags(tagsParam.split(',').map((t) => t.trim()).filter((t) => t.length > 0))
+  : []
+ const tagMode = (url.searchParams.get('tagMode') ?? 'OR') as TagMode
+ const status = (url.searchParams.get('status') ?? 'published') as StatusFilter
+ const now = Date.now()
+
+ const allDocs = await listAllDocs(kv)
+ const visibleDocs: StoredDocument[] = []
+ for (const doc of allDocs) {
+  if (shouldHideExpiredDraft(doc, now)) {
+   await deleteDoc(kv, doc.id)
+   continue
+  }
+  visibleDocs.push(doc)
+ }
+
+ const filteredDocs = visibleDocs.filter((doc) => {
+  const score = calculateScore(doc.vote, now)
+  const isDraft = score < doc.body.length
+  if (status === 'published' && isDraft) return false
+  if (status === 'draft' && !isDraft) return false
+  return tagsMatch(doc.tags, tags, tagMode)
+ })
+
+ const documents = filteredDocs
+  .map((d) => toComputed(d, false, now))
+  .sort((a, b) => b.score - a.score)
+
+ const tagTotals: Record<string, number> = {}
+ for (const doc of visibleDocs) {
+  const score = calculateScore(doc.vote, now)
+  const isDraft = score < doc.body.length
+  if (status === 'published' && isDraft) continue
+  if (status === 'draft' && !isDraft) continue
+  for (const tag of doc.tags) {
+   const vote = doc.tagVotes[tag] ?? { position: 0, velocity: 0, timestamp: doc.created }
+   tagTotals[tag] = (tagTotals[tag] ?? 0) + calculateScore(vote, now)
+  }
+ }
+ const topTags = Object.entries(tagTotals)
+  .map(([name, score]) => ({ name, score }))
+  .filter((t) => t.score > 1)
+  .sort((a, b) => b.score - a.score)
+  .slice(0, 25)
+
+ return json({ success: true, response: { documents, tags: topTags } })
+}
+
+export const onRequestPost: PagesFunction<Env> = async (context) => {
+ const url = new URL(context.request.url)
+ const kvUrl = url.searchParams.get('kv')
+ const kv = await getKV(context, true, kvUrl ?? undefined)
+ if (!kv) {
+  return json({ error: 'not authorized' }, 401)
+ }
+
+ let body: any
+ try {
+  body = await context.request.json()
+ } catch {
+  return errorText('unable to parse incoming JSON post body', 400)
+ }
+
+ const action = body?.action
+ const now = Date.now()
+
+ if (action === 'create') {
+  const title = typeof body?.title === 'string' ? body.title.trim() : ''
+  const docBody = typeof body?.body === 'string' ? body.body : ''
+  const tags = normalizeTags(asStringArray(body?.tags))
+  if (title.length === 0) return errorText('missing title', 400)
+
+  const id = safeId()
+  const tagVotes: Record<string, VoteData> = {}
+  for (const tag of tags) {
+   tagVotes[tag] = { position: 0, velocity: 0, timestamp: now }
+  }
+  const doc: StoredDocument = {
+   id,
+   title,
+   body: docBody,
+   tags,
+   created: now,
+   modified: now,
+   rev: 1,
+   vote: { position: 0, velocity: 0, timestamp: now },
+   tagVotes,
+  }
+  await insertDoc(kv, doc)
+  return json({ success: true, response: { document: toComputed(doc, true, now) } })
+ }
+
+ if (action === 'save') {
```

Comment:
We need to prevent edits to published documents here, right  now it's unprotected